pysnooper/tracer.py

```diff
@@ -19,10 +19,16 @@
 
 
 def get_local_reprs(frame, watch=()):
-    result = {key: utils.get_shortish_repr(value) for key, value
-                                                     in frame.f_locals.items()}
+    result = collections.OrderedDict(
+        (key, utils.get_shortish_repr(frame.f_locals[key]))
+        for key in frame.f_code.co_varnames if key in frame.f_locals
+    )
+
+    result.update(sorted((key, utils.get_shortish_repr(frame.f_locals[key]))
+                  for key in set(frame.f_locals) - set(frame.f_code.co_varnames)))
```

Comment:
How about: 

```python
keys = [
    key for key in frame.f_code.co_varnames if key in frame.f_locals
] + sorted(set(frame.f_locals) - set(frame.f_code.co_varnames))

result = collections.OrderedDict(
    (key, utils.get_shortish_repr(frame.f_locals[key]))
    for key in keys
)
```

Reply:
Both of your attempts are ugly in my opinion. Let me try...

Reply:
It looks like `sorted(set(frame.f_locals) - set(frame.f_code.co_varnames))` can be replaced by `frame.f_code.co_freevars`:

```python
def foo():
    y = 2
    x = 1

    def bar():
        print(y, x)

    print(bar.__code__.co_freevars)

foo()
```

But I don't know for sure.

Reply:
I wrote my own version in `9a2deda7fd63a579302989dfdb593836f8a1a1eb`, please cherrypick that. The code for getting the variables in the right order is still ugly, but now the ugliness is confined to the sorting. This way it's easier for people to understand what the actual items are without delving into the sorting code.

Reply:
Done, https://github.com/cool-RR/PySnooper/pull/113/commits/a2d848245ec2f7b4328e98d8e53ee25737c235a6.

Also I specified the the order explicitly:
    
    vars_order = frame.f_code.co_varnames + frame.f_code.co_cellvars + frame.f_code.co_freevars
 
Also, if we don't know how to sort the string, we can sort it alphabetically instead of making the tuple from dict keys. This makes code a bit simpler, faster, and the result is more predictable between python implementations. Actually, on cpython there shouldn't be any variables other than in co_varnames, co_cellvars or co_freevars